packages/replayio/src/utils/recordings/upload/uploadRecording.ts

```diff
@@ -0,0 +1,199 @@
+      recordingPath: path,
+    });
+
+    await endRecordingMultipartUpload(client, { partIds, recordingId, uploadId });
+  } else {
+    const { recordingId, uploadLink } = await beginRecordingUpload(client, {
+      buildId: buildId,
+      recordingId: id,
+      recordingSize: size,
+    });
+
+    updateRecordingLog({
+      id: id,
+      kind: RECORDING_LOG_KIND.uploadStarted,
+      recordingId: id,
+      server: replayServer,
+      timestamp: Date.now(),
+    });
+
+    await retryWithExponentialBackoff(
+      () => setRecordingMetadata(client, { metadata, recordingData }),
+      (error: unknown, attemptNumber: number) => {
+        debug(`Attempt ${attemptNumber} to set metadata failed:\n%j`, error);
+      }
+    );
+
+    await retryWithExponentialBackoff(
+      () =>
+        uploadRecordingFile({
+          recordingPath: path,
+          size,
+          uploadLink,
+        }),
+      (error: unknown, attemptNumber: number) => {
+        debug(`Attempt ${attemptNumber} to upload failed:\n%j`, error);
+      }
+    );
+
+    await endRecordingUpload(client, { recordingId });
+  }
+
+  debug("Uploaded %d bytes for recording %s", size, recording.id);
+
+  updateRecordingLog({
+    id: recording.id,
+    kind: RECORDING_LOG_KIND.uploadFinished,
+    recordingId: recording.id,
+    server: replayServer,
+    timestamp: Date.now(),
+  });
+
+  recording.uploadStatus = "uploaded";
+}
+
+async function uploadRecordingFile({
+  recordingPath,
+  size,
+  uploadLink,
+}: {
+  recordingPath: string;
+  size: number;
+  uploadLink: string;
+}) {
+  const file = createReadStream(recordingPath);
+  const resp = await fetch(uploadLink, {
+    body: file,
+    headers: { "Content-Length": size.toString(), "User-Agent": getUserAgent() },
+    method: "PUT",
+  });
+
+  if (resp.status !== 200) {
+    throw new Error(`Failed to upload recording. Response was ${resp.status} ${resp.statusText}`);
+  }
+}
+
+async function uploadRecordingFileInParts({
+  chunkSize,
+  partLinks,
+  recordingPath,
+}: {
+  chunkSize: number;
+  partLinks: string[];
+  recordingPath: string;
+}): Promise<string[]> {
+  const { size: totalSize } = statSync(recordingPath);
+
+  const results = await promiseMap<string, string>(
+    partLinks,
+    async (url: string, index: number) => {
+      return retryWithLinearBackoff(
+        async () => {
+          const partNumber = index + 1;
+          const start = index * chunkSize;
+          const end = Math.min(start + chunkSize, totalSize) - 1; // -1 because end is inclusive
+
+          debug("Uploading part %o", {
+            partNumber,
+            start,
+            end,
+            totalSize,
+            chunkSize,
+          });
+          return uploadPart(url, { recordingPath, start, end }, end - start + 1);
+        },
+        error => {
+          debug(`Failed to upload part ${index + 1}. Will be retried: %o`, error);
+        },
+        10
+      );
+    },
+    { concurrency: 10 }
```

Comment:
the legacy CLI was also setting some concurrency limits within `uploadAllRecordings`. If I'm not mistaken we lost that in the new implementation. I'm not sure what a threshold for problems is but perhaps this is something to address at some point

Reply:
PRO-83